### ISSUE TITLE
multiple languages in one file

### ISSUE THREAD

**Issue by aspell-helper** [GNUAspell/aspell#448]
**Stevan White *<stevan_white\@sf>* created a feature request on 2008-10-10 19:33:47 UTC**
*(Orig. from https://sourceforge.net/p/aspell/feature-requests/63)*

Many people would like to spell-check documents written in mixed languages \(such as English and German\).  There are work-arounds involving concatenating dictionaries, but this is a mess.

It would be nice to have a multiple-language override for the dictionary selection, say, something like
\--lang de,en
This would mean to look each word up in the German dictionary first, then in the English.  If the word is not found in either, a list of alternatives from the two languages could be presented.

A little more elaborate:  Formats such as HTML can specify the language of different parts of a document.  It would be cool if aspell could check spelling in the correct language for each part.

I have often needed this kind of functionality in several programs, and see that many other people want it as well.

Thanks\!

\------------------------------------------

Here are people asking for something similar:
http://ubuntuforums.org/showthread.php?t=896608
http://www.zimbra.com/forums/i18n-l10n-translations/9972-aspell-two-languages.html
http://osdir.com/ml/gnu.aspell.user/2003-04/msg00036.html
http://aspell.net/man-html/Multiple-Scripts.html

Here is a guy who rigged something up for Psi:
http://listas.apesol.org/pipermail/psi-devel-affinix.com/2007-October/007377.html

Here are instructions on how to concatenate dictionaries:
http://wiki.archlinux.org/index.php/User:Georgek

**Comment by pauloney** [comment]
Spelling with a "union" of two languages is a completely trivial matter -- it is the same as spelling in one language and then on another with the commands:

    cat Zanzibar.txt | aspell --lang=en list | aspell --lang=sw list | sort -u

as described in details [here](https://pseudoscripter.wordpress.com/2011/05/11/aspell-using-more-than-one-language/). 

The implementation of the feature could be as trivial as:

     --lang de,en 

being equivalent to:
 
      aspell --lang=de list | aspell --lang=en list

**Comment by maxnikulin** [comment]
> Spelling with a "union" of two languages is a completely trivial matter -- it is the same as spelling in one language and then on another with the commands:
> ```
> cat Zanzibar.txt | aspell --lang=en list | aspell --lang=sw list | sort -u
> ```

Unfortunately things are more complicated:

```
echo "program programm программа програма" | aspell --lang en list | aspell --lang ru list
programm
```

My expectation is
```
programm
програма
```
To confirm that it is not a problem of configuration:
```
echo "программа програма" | aspell --lang ru list
програма
```

If it is necessary to get list of suggestions for correction than it does not work at all
```
echo "program programm программа програма" | aspell -d en -a | aspell -d ru -a
@(#) International Ispell Version 3.1.20 (but really Aspell 0.60.8)
Error: The word "(#) International Ispell Version 3.1.20 (but really Aspell 0.60.8)" is invalid. The character '(' (U+28) may not appear at the beginning of a word.
Error: The word "" is invalid. Empty string.
Error: The word "programm 5 8: program, programs, programmer, programmed, program's" is invalid. The character ' ' (U+20) may not appear in the middle of a word.
```
While both languages works well in isolation
```
echo "program programm" | aspell --lang en -a
@(#) International Ispell Version 3.1.20 (but really Aspell 0.60.8)
*
& programm 5 8: program, programs, programmer, programmed, program's

echo "программа програма" | aspell --lang ru -a
@(#) International Ispell Version 3.1.20 (but really Aspell 0.60.8)
*
& програма 13 10: программа, программ, программе, программу, программы, программах, программам, программка, параграмма, программою, проиграна, параграмм, погрома
```

Two languages works to some extent in hunspell. My first impression was however that it can provide suggested correction to from the first dictionary. Actually its behavior is more funny
```
echo "program programm программа програма" | hunspell -d en_US,ru_RU
Hunspell 1.7.0
*
& programm 4 8: program, programs, program m, mammogram
*
& програма 3 27: программа, программ, барограмма

echo "програма программа programm program" | hunspell -d en_US,ru_RU
Hunspell 1.7.0
# програма 0
*
# programm 19
*
```

After looking into some recipes how to achieve spell checking of mixed-language text, my impression is that not all people realize that the following cases may differ
- Additional dictionaries for the same language, e.g. medical or jargon for English.
- Same language having more than one script http://aspell.net/0.60.7/man-html/Multiple-Scripts.html
- Different languages that may have common subset of letters or may be completely distinct, allowing language detection by character code points.

P.S. The following command works in vim for more than a decade (highlight of misspelled words, suggestions. It requires to download a dictionary in a special format however), so I am really surprised that spell checkers still have problems with multi-lingual texts.

    :set spelllang=en,ru spell

**Comment by novoid** [comment]
@maxnikulin wrote:

> P.S. The following command works in vim for more than a decade (highlight of misspelled words, suggestions. It requires to download a dictionary in a special format however), so I am really surprised that spell checkers still have problems with multi-lingual texts.
> 
> ```
> :set spelllang=en,ru spell
> ```

Interesting. Thanks for this idea.

The only thing that puzzles me here: when I'm checking for EN/DE in parallel and I do find a correctly written word that gets highlighted which I want to add to my personal wordlist: to what personal wordlist is this added to? I do have `~/.aspell.de.pws` and `~/.aspell.en.pws` as two different candidates for that.

**Comment by maxnikulin** [comment]
> The only thing that puzzles me here: when I'm checking for EN/DE in parallel and I do find a correctly written word that gets highlighted which I want to add to my personal wordlist: to what personal wordlist is this added to? I do have `~/.aspell.de.pws` and `~/.aspell.en.pws` as two different candidates for that.

By default it is added to the dictionary specified first: http://vimdoc.sourceforge.net/htmldoc/spell.html#zg However one can use a custom list of word lists (languages, personal common, for particular project, etc), see [spellfile option](http://vimdoc.sourceforge.net/htmldoc/options.html#'spellfile'), and add a number before the command to add a word. Maybe it is not really convenient, but it is at least viable when no advanced heuristics is implemented (that will fail anyway for some cases). Necessity to remember numbers of word lists is not user friendly.

For aspell there is an interesting idea to switch language in response to a command sent through pipe: https://github.com/GNUAspell/aspell/issues/176

I am thinking on widely suggested workaround to combine 2 languages in single dictionary. I have not realized yet what features I will miss. I do not like idea to merge just word lists. Things becomes more tricky with affixes. Besides exclude rules are necessary when characters are distinct, at least the file for Russian language occupies whole alphabet leaving no room for others:

    grep '^SFX' /usr/lib/aspell/ru_affix.dat | cut -c 1-5 | sort -u | wc -l
    26

Another approach is multiplexing wrapper that sends every word to real per-language aspell processes and somehow combines responses. Parsing command line options will be a mess since some of them should be passed for specific language only. Options like `-Wl,...` in gcc may be a rescue, however it breaks compatibility with ispell. Since aspell has 8-bit internals, mixing languages withing single process (instead of multiplexing proxy) may cause some complications.

I think, this issue deserves to be linked from related TODO item at the main page of the Aspell web site.

**Comment by kevina** [comment]
@maxnikulin @novoid how do you use aspell (via vim, emacs, standalone, etc.) and do you care about suggestions when a word is misspelled?

**Comment by novoid** [comment]
> @maxnikulin @novoid how do you use aspell (via vim, emacs, standalone, etc.) and do you care about suggestions when a word is misspelled?
 
I'm afraid I don't understand your question.

My vim uses aspell out-of-the-box I guess and provides suggestions on `z =`. Sorry, I did my vim setup over two decades ago and did not change much in the meantime.

My emacs ([config here](https://github.com/novoid/dot-emacs/blob/master/config.org)) is using aspell out-of-the-box (you can manually override `ispell-program-name`) after loading `(require 'ispell)`. Following lines map F7 to go to the next error and F8 top open the suggestion menu:

```
(define-key global-map [(f7)] 'flyspell-goto-next-error)
(define-key global-map [(f8)] 'flyspell-correct-word-before-point)
```

Did this answer your question?

**Comment by maxnikulin** [comment]
> how do you use aspell (via vim, emacs, standalone, etc.) and do you care about suggestions when a word is misspelled?

I am unsure, but I do not see apparent dependency of Vim system packages on external spell checkers. So I believe I use built-in one and (that is why) it just works with mixed-language files. Thus I do not use aspell with vim. I found this feature request when I tried to achieve similar behavior in Emacs.

I rarely use a command to check whole file in the interactive mode. Usually I just correct highlighted words with regular editing commands, anyway reading list of variants requires some attention. Sometimes I ask for suggestions (`z=` if I have not recognized a typo fast enough). I have realized that in some cases it might be useful to see alternatives of a word that is not marked as wrong. It works in Vim, but Emacs just reports that the word is correct in response to M-$. In Vim enabling spell check (explicitly or through a setting for specific file type) immediately highlights typos, commands like M-x flyspell-buffer are not necessary. I never used ispell-complete-word or similar commands for prose. I do not think it is useful for Russian due to variations of final part of words.

**Comment by david-vicente** [comment]
+1 to this feature

**Comment by dehidehidehi** [comment]
+1
